### PR TITLE
Add bindings() option to collect bindings generated by bindingsStream/quadStream

### LINKED ISSUE #635: Add option to produce non-streaming results
#### Issue type:

- :heavy_plus_sign: Feature request

____
#### Description:

For many use cases, the streaming capabilities of Comunica are irrelevant.
For instance, the following is a common pattern:
```javascript
const array = [];
result.bindingsStream.on('data', (data) => {
  array.push(data);
});
result.bindingsStream.on('end', () => {
  // handle array
});
```

We could additionally introduce something like this for convenience:

```
const array = await result.bindings();
```

This should also be done for quadStreams.

### PR DESCRIPTION
**Pull request by FlorianFV**
Closes #635

### CODE REVIEW

File: packages/actor-init-sparql/lib/ActorInitSparql-browser.ts [hunk 1]
```diff
@@ -169,6 +169,32 @@ export class ActorInitSparql extends ActorInit implements IActorInitSparqlArgs {
     // Execute query
     const resolve: IActionQueryOperation = { context, operation };
     const output = await this.mediatorQueryOperation.mediate(resolve);
+    // Set bindings
```
Comment (rubensworks) [inline packages/actor-init-sparql/lib/ActorInitSparql-browser.ts:172]: I would suggest moving all of this new code into a new static method (`enhanceQueryResults`?). This would also simplify testing.

File: packages/actor-init-sparql/lib/ActorInitSparql-browser.ts [hunk 2]
```diff
@@ -169,6 +169,32 @@ export class ActorInitSparql extends ActorInit implements IActorInitSparqlArgs {
...
+    if ((<any> output).bindingsStream) {
```
Comment (rubensworks) [inline packages/actor-init-sparql/lib/ActorInitSparql-browser.ts:173]: Casting to `any` should be avoided if possible. Does the compiler accept `'bindingsStream' in output` instead?

File: packages/actor-init-sparql/lib/ActorInitSparql-browser.ts [hunk 3]
```diff
@@ -169,6 +169,32 @@ export class ActorInitSparql extends ActorInit implements IActorInitSparqlArgs {
...
+      (<IActorQueryOperationOutputBindings> output).bindings = () => {
+        const result: any[] = [];
+        (<IActorQueryOperationOutputBindings> output).bindingsStream.on('data', data => {
+          result.push(data.toObject());
```
Comment (rubensworks) [inline packages/actor-init-sparql/lib/ActorInitSparql-browser.ts:177]: No need to convert to object.

File: packages/actor-init-sparql/lib/ActorInitSparql-browser.ts [hunk 4]
```diff
@@ -169,6 +169,32 @@ export class ActorInitSparql extends ActorInit implements IActorInitSparqlArgs {
     // Execute query
     const resolve: IActionQueryOperation = { context, operation };
     const output = await this.mediatorQueryOperation.mediate(resolve);
+    // Set bindings
+    if ((<any> output).bindingsStream) {
+      (<IActorQueryOperationOutputBindings> output).bindings = () => {
+        const result: any[] = [];
```
Comment (rubensworks) [inline packages/actor-init-sparql/lib/ActorInitSparql-browser.ts:175]: Result type can be made more concrete.

File: packages/actor-init-sparql/lib/ActorInitSparql-browser.ts [hunk 5]
```diff
@@ -169,6 +169,32 @@ export class ActorInitSparql extends ActorInit implements IActorInitSparqlArgs {
...
+        });
+        return new Promise((resolveP, reject) => {
+          (<IActorQueryOperationOutputBindings> output).bindingsStream.on('end', () => {
+            resolveP(result);
+          });
+        });
+      };
+    } else if ((<any> output).quadStream) {
+      (<IActorQueryOperationOutputQuads> output).bindings = () => {
+        const result: any[] = [];
```
Comment (rubensworks) [inline packages/actor-init-sparql/lib/ActorInitSparql-browser.ts:187]: Result type can be made more concrete.

File: packages/actor-init-sparql/lib/ActorInitSparql-browser.ts [hunk 6]
```diff
@@ -169,6 +169,32 @@ export class ActorInitSparql extends ActorInit implements IActorInitSparqlArgs {
...
     const resolve: IActionQueryOperation = { context, operation };
     const output = await this.mediatorQueryOperation.mediate(resolve);
+    // Set bindings
+    if ((<any> output).bindingsStream) {
+      (<IActorQueryOperationOutputBindings> output).bindings = () => {
+        const result: any[] = [];
+        (<IActorQueryOperationOutputBindings> output).bindingsStream.on('data', data => {
+          result.push(data.toObject());
+        });
+        return new Promise((resolveP, reject) => {
```
Comment (rubensworks) [inline packages/actor-init-sparql/lib/ActorInitSparql-browser.ts:179]:
```suggestion
        return new Promise((resolve, reject) => {
```

File: packages/actor-init-sparql/lib/ActorInitSparql-browser.ts [hunk 7]
```diff
@@ -169,6 +169,32 @@ export class ActorInitSparql extends ActorInit implements IActorInitSparqlArgs {
...
+          });
+        });
+      };
+    } else if ((<any> output).quadStream) {
+      (<IActorQueryOperationOutputQuads> output).bindings = () => {
+        const result: any[] = [];
+        (<IActorQueryOperationOutputQuads> output).quadStream.on('data', data => {
+          result.push(data);
+        });
+        return new Promise((resolveP, reject) => {
```
Comment (rubensworks) [inline packages/actor-init-sparql/lib/ActorInitSparql-browser.ts:191]:
```suggestion
        return new Promise((resolve, reject) => {
```

File: packages/actor-init-sparql/lib/ActorInitSparql-browser.ts (hunk 6 above)
Comment (rubensworks) [reply packages/actor-init-sparql/lib/ActorInitSparql-browser.ts:179]: Also, reject is never called. You should also listen to error events and forward those.

File: packages/actor-init-sparql/lib/ActorInitSparql-browser.ts (hunk 7 above)
Comment (rubensworks) [reply packages/actor-init-sparql/lib/ActorInitSparql-browser.ts:191]: Also, reject is never called. You should also listen to error events and forward those.

File: packages/bus-query-operation/lib/ActorQueryOperation.ts [hunk 8]
```diff
@@ -249,6 +249,10 @@ export interface IActorQueryOperationOutputBindings extends IActorQueryOperation
    * The stream of bindings resulting from the given operation.
    */
   bindingsStream: BindingsStream;
+  /**
+   * The collection of bindings after an 'end' event occured.
+   */
+  bindings?: () => Promise<any[]>;
```
Comment (rubensworks) [inline packages/bus-query-operation/lib/ActorQueryOperation.ts:255]: Result type should be more concrete.

File: packages/bus-query-operation/lib/ActorQueryOperation.ts [hunk 9]
```diff
@@ -269,6 +273,10 @@ export interface IActorQueryOperationOutputQuads extends IActorQueryOperationOut
    * The stream of quads.
    */
   quadStream: RDF.Stream & AsyncIterator<RDF.Quad>;
+  /**
+   * The collection of bindings after an 'end' event occured.
+   */
+  bindings?: () => Promise<any[]>;
```
Comment (rubensworks) [inline packages/bus-query-operation/lib/ActorQueryOperation.ts:279]: Result type should be more concrete.
Also, `quads`?

File: packages/bus-query-operation/lib/ActorQueryOperation.ts [hunk 10]
```diff
@@ -249,6 +249,10 @@ export interface IActorQueryOperationOutputBindings extends IActorQueryOperation
    * The stream of bindings resulting from the given operation.
    */
   bindingsStream: BindingsStream;
+  /**
```
Comment (rubensworks) [inline packages/bus-query-operation/lib/ActorQueryOperation.ts:252]: These new fields being optional makes them very inconvenient to use.

In the actor-init-sparql package, I would suggest creating the following instead:
* IQueryResultBindings: extends from IActorQueryOperationOutputBindings, and adds this new bindings method.
* IQueryResultQuads: extends from IActorQueryOperationOutputQuads, and adds this new quads method.
* IQueryResultBoolean: extends from IActorQueryOperationOutputBoolean, and adds nothing.
* IQueryResult: _union type_ over the three above.

The return type of `query` in ActorInitSparql can then be changed to return `IQueryResult`.
Then the "optional" any "any" stuff won't be needed anymore.

File: packages/actor-init-sparql/test/ActorInitSparql-test.ts [hunk 11]
```diff
@@ -583,6 +584,13 @@ describe('ActorInitSparql', () => {
       });
     });
 
+    it('bindings() should collect all bindings until "end" event occurs on triples', async() => {
```
Comment (rubensworks) [inline packages/actor-init-sparql/test/ActorInitSparql-test.ts:587]: The tests look good, but will probably have to be changed a bit after my suggestions.

Could you also add some tests for empty results and erroring results?

File: packages/actor-init-sparql/lib/ActorInitSparql-browser.ts [hunk 12]
```diff
@@ -84,6 +87,38 @@ export class ActorInitSparql extends ActorInit implements IActorInitSparqlArgs {
     super(args);
   }
 
+  public static enhanceQueryResults(results: IQueryResult): IQueryResult {
```
Comment (rubensworks) [inline packages/actor-init-sparql/lib/ActorInitSparql-browser.ts:90]: Input here should be of type `IActorQueryOperationOutput`, output type is correct.

Comment (rubensworks) [inline packages/actor-init-sparql/lib/ActorInitSparql-browser.ts:90]: Add tsdoc comment above, such as _Add convenience methods to query results_, with a description of the param (and return type).

File: packages/bus-query-operation/lib/ActorQueryOperation.ts [hunk 13]
```diff
@@ -237,7 +237,7 @@ export function getMetadata(actionOutput: IActorQueryOperationOutputStream): Pro
 }
 
 /**
- * Query operation output for a bindings stream.
+ * Query operation output for a bindings stream. (without bindings operator)
```
Comment (rubensworks) [inline packages/bus-query-operation/lib/ActorQueryOperation.ts:240]: The comment changes in this file are not really needed, I think they make the description too complex.
